File: 704-binary-search.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        // Time: O(logN)
        // Space: O(1)
        
        int left = 0;
        int right = nums.size() - 1;
        int mid;
        sort(nums.begin(), nums.end());

        while (left <= right) {
            mid = left + (right - left)/2;
            if (target < nums[mid])
                right = mid - 1;
            else if (target > nums[mid])
                left = mid + 1;
            else
                return mid;
        }

        return -1;
    }
};

File: 0700-0799/704-binary-search.py
# https://leetcode.com/problems/binary-search/description/

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1

        while left < right:
            # 1 2 3 4
            # 1 2 2 2 2 2 3
            mid = left + (right - left) // 2
            # mid = right - (right - left) // 2
            # if nums[mid] == target:
            #     return mid
            if nums[mid] >= target:
                right = mid
            else:
                left = mid + 1

        return left if nums[left] == target else -1
